eip/EIPS/eip-aa-rules.md

```diff
@@ -0,0 +1,351 @@
+
+### Alt-mempools rules:
+
+Alternate mempool is an agreed-upon rule that the bundlers may opt in to.
+
+* **[ALT-010]** The bundler listens to the alt-mempool "topic" over the P2P protocol
+* **[ALT-020]** The alt mempool rules MUST be checked only when a canonical rule is violated
+    * That is, if validation follows the canonical rules above, it is not considered part of an alt-mempool.
+* **[ALT-030]** Bundlers SHOULD forward `UserOperations` to other bundlers only once, regardless of how many (shared) alt-mempools they have. \
+The receiving bundler validates the `UserOperations`, and based on the above rules (and subscribed alt-mempools) decides for which alt-mempools to "tag" it.
+
+
+## Security Considerations
+
+### Possible Attacks
+
+Below is a list of possible attacks that were considered and a reference to the above rule that prevents it.
+
+This list does not attempt to be an exhaustive list of attacks.
+These attacks are examples provided to describe and rationalise the reason for the above rules.
+
+### Sample Known (expensive) attack:
+
+* This attack can't be fully mitigated by the above validation rules.
+  But the rules made sure it is expensive to maintain.
+* Deploy a paymaster which has the following line in the validation function: \
+`require((perSenderMap[userOp.sender]&1)==0);`
+* Submit a large number of `UserOperations` with different senders into the mempool.
+* Front-run the mempool, and modify the `perSenderMap` for each sender.
+* By the time the `UserOperations` are validated the second time they are all invalid.
+* The cost of first run is 20000 gas per account, but the second run is 5000 gas per account.
+* NOTE: Because of the rule **[STO-021]** "forbid associated storage during creation", the accounts have to be pre-deployed. This introduces a one-time cost of ~40000 gas per account.
+* Calculate the cost of blocking the network:
+* Assuming a bundler can process 1000 `UserOperations`/second
+* Polygon - (2-second blocks) need 2000 `UserOperations`
+    * The initial cost (deploy+initial state change) is 2000*60000 = 120M gas
+    * Sustained cost (only storage modification) - 2000*5000 =  10M gas
+* Mainnet- (12-second blocks) need 12000 UserOps
+    * The initial cost (deploy+initial state change) is 12000*60000 = 720M gas
+    * Sustained cost (only storage modification) - 12000*5000 =  70Mgas
+* Total attack cost
+    * Polygon (@150 gwei) - $18 + $1.5 per block ($2700/hour)
+    * Mainnet: (@40 gwei) - $51840 + $5040 per block ($1.5M/hour)
+    * Note that the above costs ignore the gas-price increase because the underlying network is also overloaded.
+* Mitigations
+    * A **staked factory:** 
+        * Can attempt such an attack once, since after invalidating a single bundle creation its reputation drops and another staked bundler has to be used.
```

Comment:
Can we secure this further by allowing reputation information also to be shared via p2p interface of the bundlers? @drortirosh @yoavw

Reply:
@kristofgazso This is problematic, reputation was always meant to be local to the bundler, otherwise bundlers could not only censor an account/paymaster/factory/aggregator, they could also direct other bundlers to censor them. Sharing reputation as part of the p2p core protocol opens a can of worms in that regard.